### PR TITLE
Add interpreter for BatchNormTrainingOp

### LINKED ISSUE #1122: Add interpreter for BatchNormTrainingOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand`: tensor of floating-point type.
(I2) `scale`: 1-dimensional tensor of floating-point type.
(I3) `offset`: 1-dimensional tensor of floating-point type.
(I4) `epsilon`: constant of type `f32`.
(I5) `feature_index`: constant of type `si64`.
(C1) 0 <= `feature_index` < rank(`operand`).
(C2) `operand`, `scale`, `offset`, `result`, `batch_mean` and `batch_var`
     have the same element type.
(C3) size(`scale`) = `dim(operand, feature_index)`.
(C4) size(`offset`) = `dim(operand, feature_index)`.
(C5) size(`batch_mean`) = `dim(operand, feature_index)`.
(C6) size(`batch_var`) = `dim(operand, feature_index)`.
(C7) `operand` and `output` have the same type.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor of floating-point type. (Covered by ODS).
I2: a) `scale` is not a 1-dimensional tensor. (Covered by ODS).
    b) `scale` is not a tensor of floating-point type. (Covered by ODS).
I3: a) `offset` is not a 1-dimensional tensor. (Covered by ODS).
    b) `offset` is not a tensor of floating-point type. (Covered by ODS).
I4: a) `epsilon` is not a constant of type `f32`. (Covered by ODS).
I5: a) `feature_index` is not a constant of type `si64`. (Covered by ODS).
C1: a) `feature_index` < 0.
    b) `feature_index` >= rank(`operand`).
C2: a) element_type(`operand`) != element_type(`scale`). (Covered by ODS).
    b) element_type(`operand`) != element_type(`offset`). (Covered by ODS).
    c) element_type(`operand`) != element_type(`result`). (Covered by ODS).
    d) element_type(`operand`) != element_type(`batch_mean`). (Covered by ODS).
    e) element_type(`operand`) != element_type(`batch_var`). (Covered by ODS).
C3: a) size(`scale`) != dim(operand, feature_index).
C4: a) size(`offset`) != dim(operand, feature_index).
C5: a) size(`batch_mean`) != dim(operand, feature_index).
C6: a) size(`batch_var`) != dim(operand, feature_index).
C7: a) type(`operand`) != type(`output`).
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1a: `feature_index` < 0.
C1b: `feature_index` >= rank(`operand`).
C3a: size(`scale`) != dim(operand, feature_index).
C4a: size(`offset`) != dim(operand, feature_index).
C5a: size(`batch_mean`) != dim(operand, feature_index).
C6a: size(`batch_var`) != dim(operand, feature_index).
C7a: type(`operand`) != type(`output`).
```

closes #1122

### CODE REVIEW

**Comment by ghpvnist** [comment]
Please review the latest [commit](https://github.com/openxla/stablehlo/pull/1393/commits/ada527149db810369c03b85c372a9bee12d17579) only.

**Comment by burmako** [comment]
@ghpvnist Once all the diffbase PRs are merged, I'd like to review this PR to discuss how to express evalBatchNormTrainingOp doing reductions. I've also left some feedback earlier in #1394 which actually applied to this PR.

**Comment by ghpvnist** [comment]
@burmako The diffbase PRs are now all merged with comments from other PR addressed.

File: stablehlo/reference/Ops.cpp
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:106]: I do not see the point of this generic function. 
We are using this function to create a `DenseIntElementsAttr` to be passed on to `inferReduceOp`. Are we using this function is some other places as well?

File: stablehlo/reference/Ops.cpp
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:133]: delete? same below

File: stablehlo/reference/Ops.cpp
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:135]: `nextValToAdd`

File: stablehlo/reference/Ops.cpp
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:109]: It seems that you can trying to generalize this function. In the current use case, we can specialize it only for 1-D tensor for dimensions, right?

File: stablehlo/reference/Ops.cpp
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp]: Thanks!

File: stablehlo/reference/Ops.cpp
Comment (sdasgup3) [inline stablehlo/reference/Ops.cpp:106]: We can remove this right?

**Review by sdasgup3 — APPROVED** [review]
With one trivial comment

File: stablehlo/reference/Ops.cpp
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp]: Should be gone now!

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:236]: Why do we say `{results}` here if a regular `results` would suffice? Same for other locations in Ops.cpp - apologies for not noticing this in other reviews.

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:134]: Could use `operand.getAxes()` from another PR here as well.

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:130]: Can we pass `resultType` here as well? I think `resultTypes[1]` would be it, and the we wouldn't need to call `inferReduceOp`, right? (Same for `computeVariance` and `resultTypes[2]`).

Comment (burmako) [inline stablehlo/reference/Ops.cpp:130]: I don't think constness of `Axis` is buying us much here?

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:151]: `sum` to follow the name in the spec?

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:128]: Since we're partially evaluating the implementation of `evalReduceOp` anyway, we might as well partially evaluate it further and inline the call to evalAddOp:

```C++
static Tensor computeSum(const Tensor &input, const Tensor &initValue,
                         const Axes &dimensions, ShapedType resultType) {
  Tensor result(resultType);
  for (auto resultIt = result.index_begin(); resultIt != result.index_end();
       ++resultIt)
    result.set(*resultIt, initValue.get({}));

  for (auto inputIt = input.index_begin(); inputIt != input.index_end();
       ++inputIt) {
    Index resultIndex;
    for (auto [inputAxis, inputIndexElement] : llvm::enumerate(*inputIt)) {
      if (llvm::is_contained(dimensions, inputAxis)) continue;
      resultIndex.push_back(inputIndexElement);
    }
    result.set(resultIndex, result.get(resultIndex) + input.get(*inputIt));
  }
  return result;
}
```

Comment (burmako) [inline stablehlo/reference/Ops.cpp:128]: In the code snippet above, I also renamed `operand` to `input` to follow `evalReduceOp` and changed the logic of computing `resultIndex` which was changed during review of the corresponding PR.

**Review by burmako — CHANGES_REQUESTED** [review]
Only reviewed evalBatchNormTrainingOp and its call graph.

File: stablehlo/reference/Ops.cpp
Comment (burmako) [inline stablehlo/reference/Ops.cpp:122]: `Axes dimensions = operand.getAxes();`

**Review by burmako — APPROVED** [review]
LGTM modulo one minor comment

**Pull request merged**
